plugin/field_data_utils.go

```diff
@@ -0,0 +1,105 @@
+package gcpsecrets
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/errwrap"
+	"github.com/hashicorp/vault-plugin-secrets-gcp/plugin/util"
+	"github.com/hashicorp/vault/sdk/framework"
+)
+
+type inputParams struct {
+	name       string
+	secretType string
+
+	hasBindings bool
+	rawBindings string
+	bindings    ResourceBindings
+
+	project             string
+	serviceAccountEmail string
+
+	scopes []string
+}
+
+func (input *inputParams) parseOkInputSecretType(d *framework.FieldData) (warnings []string, err error) {
+	secretType := d.Get("secret_type").(string)
+	if secretType == "" && input.secretType == "" {
+		return nil, fmt.Errorf("secret_type is required")
+	}
+	if input.secretType != "" && secretType != "" && input.secretType != secretType {
+		return nil, fmt.Errorf("cannot update secret_type - currently %q", input.secretType)
+	}
+
+	switch secretType {
+	case SecretTypeKey, SecretTypeAccessToken:
+		input.secretType = secretType
+		return nil, nil
+	default:
+		return nil, fmt.Errorf("invalid secret_type %q", secretType)
+	}
+}
+
+func (input *inputParams) parseOkInputServiceAccountEmail(d *framework.FieldData) (warnings []string, err error) {
+	email := d.Get("service_account_email").(string)
+	if email == "" && input.serviceAccountEmail == "" {
+		return nil, fmt.Errorf("email is required")
+	}
+	if input.serviceAccountEmail != "" && email != "" && input.serviceAccountEmail != email {
```

Comment:
This check here is pretty essential on how this method behaves and how it should be called. Can we drop a comment here (or better yet, a  godoc on the func) on this behavior?

Reply:
Added a godoc to explain this.